Allow distinct clearing price per lease duration bucket

guggero opened the pull request:
This PR contains the client part of the change to switch over to distinct lease duration buckets for batches.

UPDATE: Made this non-breaking. Clients with batch version `0` won't be able to submit orders with a lease duration != 2016.

UPDATE2: Depends on #193 which was split off to make the server changes easier to split into multiple PRs.

Roasbeef commented:
I think this is ready for review now?

guggero commented:
Yes, has been for a while.

halseth commented on poolrpc/auctioneer.proto:
```diff
@@ -956,6 +956,32 @@ message ServerOrderStateResponse {
+    CLEARING_MARKET indicates that we're accepting orders, and fully clearing
+    the market for this duration.
+    */
+    CLEARING_MARKET = 3;
```
Call this `market open`?

halseth commented on poolrpc/auctioneer.proto:
```diff
@@ -441,48 +441,56 @@ message OrderMatchPrepare {
     server when executing the orders.
     */
-    repeated AccountDiff charged_accounts = 3;
+    repeated AccountDiff charged_accounts = 2;
```
I think we should avoid shifting the protos here. Feels like it could lead to some unexpected issues.

guggero replied:
Yeah, made all changes non-breaking for older clients.

halseth commented on clientdb/batch_snapshot.go:
```diff
@@ -426,13 +438,31 @@ func serializeLocalBatchSnapshot(w io.Writer, b *LocalBatchSnapshot) error {
 		}
 	}
 
+	// New batch versions have an additional map of duration->price that we
```
Great solution 👍

halseth commented on rpcserver.go:
```diff
@@ -1810,10 +1810,19 @@ func (s *rpcServer) prepareLeasesResponse(ctx context.Context,
+				// duration, except for older batch versions
+				// where there was just one single duration.
+				duration := bidDuration
+				if batch.Version != order.DefaultVersion {
```
I don't understand why this is needed. Won't the bid duration always match the duration for the batch it was matched in?

guggero replied:
This is for legacy legacy orders where the durations were defined free style. When de-serializing those batch snapshots, the single clearing price is put into the map with the 2016 duration as the key. Otherwise we wouldn't know whether to look it up using the bid or ask duration (as they could differ before the uniform 2016 duration was introduced).

halseth commented on poolrpc/auctioneer.proto:
```diff
@@ -965,21 +991,18 @@ message TermsResponse {
+    the duration buckets currently are in.
     */
-    map<uint32, bool> lease_durations = 4;
+    map<uint32, DurationBucketState> lease_duration_buckets = 4;
```
wouldn't it be pretty easy to not break the API by adding this as a new field?

halseth reviewed (changes requested):
First pass. Main question is whether we can avoid breaking the API.

guggero commented:
I was able to make everything non-breaking!
Still need to do some manual tests to confirm, but I think this PR is ready for a second look.

wpaulino commented on clientdb/batch_snapshot.go:
```diff
@@ -465,6 +495,40 @@ func deserializeLocalBatchSnapshot(r io.Reader) (*LocalBatchSnapshot, error) {
+	// If it wasn't found in the database, then we'll assume the default
+	// value.
+	case io.EOF, io.ErrUnexpectedEOF:
+		b.ClearingPrices[order.LegacyLeaseDurationBucket] = clearingPrice
```
Could avoid the use of `EOF` markers by instead checking if the clearing price is non-zero signaling an old batch.

guggero replied:
Good idea! I thought it might screw with our unit tests if we always write a zero price. But it turns out this works well.

wpaulino commented on rpcserver.go:
```diff
@@ -239,6 +242,14 @@ func (s *rpcServer) consumePendingOpenChannels(
+
+			// If the lnd node shut down, there's no use continuing.
+			if err == io.EOF || err == io.ErrUnexpectedEOF ||
+				status.Code(err) == codes.Unavailable {
```
:+1:

wpaulino reviewed (commented):
Non-breaking changes look good on first pass!

halseth commented on poolrpc/trader.proto:
```diff
@@ -871,7 +871,11 @@ message LeaseDurationRequest {
+    The set of lease durations the market is currently accepting and the state
+    the duration buckets currently are in.
+    */
+    map<uint32, DurationBucketState> lease_duration_buckets = 1;
```
this breaks clients of the trader service.

guggero replied:
You're right, maybe someone is already using the RPC or REST interface. Deprecated the legacy format instead of replacing it.

halseth commented on poolrpc/auctioneer.proto:
```diff
@@ -483,6 +498,12 @@ message OrderMatchPrepare {
     uint32 batch_version = 9;
+
+    /*
+    Maps the distinct lease duration markets to the orders that were matched
```
Nice! 👍

halseth commented on order/rpc_parse.go:
```diff
@@ -274,7 +296,6 @@ func ParseRPCBatch(prepareMsg *poolrpc.OrderMatchPrepare) (*Batch,
 	}
 
 	// Convert clearing price, fee rate and rebate.
-	b.ClearingPrice = FixedRatePremium(prepareMsg.ClearingPriceRate)
```
should be removeed in later commit?

halseth commented on clientdb/batch_snapshot.go:
```diff
@@ -382,8 +383,19 @@ func getSnapshotBuckets(tx *bbolt.Tx) (*bbolt.Bucket, *bbolt.Bucket,
+	// extract it from the map which will only contain one entry for that
+	// older version. For new batch versions we just store a zero price.
+	var clearingPrice order.FixedRatePremium
+	if len(b.ClearingPrices) == 1 {
```
keep in mind that this might have only one element for new version as well, if only one bucket had matches.

guggero replied:
Removed, based on Wilmer's suggestion above.

halseth reviewed (changes requested):
Great work on this PR! The latest version reads well to me. 

Also needs a rebase!

wpaulino commented on clientdb/batch_snapshot_test.go:
```diff
@@ -112,44 +108,91 @@ var testSnapshot = &LocalBatchSnapshot{
+				t.Fatal(err)
+			}
+
+			if !reflect.DeepEqual(pre, post) {
```
Nit: Is `require.Equal` any better here?

guggero replied:
Yes, the diff seems to be much cleaner :+1:

halseth reviewed (approved):
Awesome, this turned out very clean in the end 👌 

Only nit from me is that I think we should make deprecated fields more clear by removing all but the deprecated message from their doc.

guggero commented on poolrpc/auctioneer.proto:
```diff
@@ -979,6 +1005,7 @@ message TermsResponse {
 
     /*
     The maximum order duration in blocks currently allowed by the auctioneer.
+    Deprecated, use explicit order duration from lease_duration_buckets.
```
Good idea, done :+1:

guggero commented:
Rebased and addressed all comments.

Even though this is ready to be merged now, I think we should wait until the server part is close to being merged as well.

halseth commented on clientdb/batch_snapshot.go:
```diff
@@ -481,6 +504,34 @@ func deserializeLocalBatchSnapshot(r io.Reader) (*LocalBatchSnapshot, error) {
+	}
+
+	for i := uint32(0); i < numPrices; i++ {
+		err = ReadElements(r, &duration, &price)
```
this is probably okay since these are just `uint32`, but triggers my spidey sense :P Could define these in the inner scope to make clear that each iteration is not overwriting the previous value.

halseth reviewed (approved):
Re-ACK ✅

guggero commented on clientdb/batch_snapshot.go:
```diff
@@ -465,6 +488,34 @@ func deserializeLocalBatchSnapshot(r io.Reader) (*LocalBatchSnapshot, error) {
+	// need to read any further.
+	if clearingPrice > 0 {
+		b.ClearingPrices[order.LegacyLeaseDurationBucket] = clearingPrice
+
```
Hmm, I thought I already replied to this, but apparently that comment got lost. I added the newline on purpose to make the return stand out a bit more since it's an early exit from the function.

guggero commented:
Replaced by #199.
